File: packages/node-sdk/src/resource-attributes.ts
import os from "node:os";
import { execFileSync } from "node:child_process";

export type ResourceAttributesInput =
  | Record<string, unknown>
  | (() => Record<string, unknown>);

let cachedGitSha: string | null | undefined;

function detectGitSha(): string | null {
  if (cachedGitSha !== undefined) return cachedGitSha;

  const envSha =
    process.env.OBYFLOW_GIT_SHA ||
    process.env.GIT_SHA ||
    process.env.GIT_COMMIT ||
    process.env.GITHUB_SHA ||
    process.env.VERCEL_GIT_COMMIT_SHA ||
    process.env.HEROKU_SLUG_COMMIT ||
    null;

  if (envSha) {
    cachedGitSha = envSha;
    return cachedGitSha;
  }

  try {
    const sha = execFileSync("git", ["rev-parse", "HEAD"], {
      cwd: process.cwd(),
      stdio: ["ignore", "pipe", "ignore"],
    })
      .toString()
      .trim();
    cachedGitSha = sha || null;
  } catch {
    cachedGitSha = null;
  }

  return cachedGitSha;
}

export function baseResourceAttributes(): Record<string, unknown> {
  const attrs: Record<string, unknown> = {
    hostname: os.hostname(),
    pid: process.pid,
    node_version: process.version,
  };
  const gitSha = detectGitSha();
  if (gitSha) attrs.git_sha = gitSha;
  return attrs;
}

export function resolveResourceAttributes(
  custom?: ResourceAttributesInput,
): Record<string, unknown> {
  const base = baseResourceAttributes();
  if (!custom) return base;
  const extra = typeof custom === "function" ? custom() : custom;
  return { ...base, ...extra };
}

export function _resetResourceAttributesCacheForTests(): void {
  cachedGitSha = undefined;
}

File: python/obyflow-python/obyflow/resource_attributes.py
from __future__ import annotations

# pylint: disable=invalid-name,global-statement

import os
import platform
import socket
import subprocess
from typing import Any, Callable, Dict, Optional, Union

ResourceAttributesInput = Union[Dict[str, Any], Callable[[], Dict[str, Any]]]

_cached_git_sha: Optional[str] = None
_git_sha_resolved = False

_GIT_SHA_ENV_VARS = (
    "OBYFLOW_GIT_SHA",
    "GIT_SHA",
    "GIT_COMMIT",
    "GITHUB_SHA",
    "VERCEL_GIT_COMMIT_SHA",
    "HEROKU_SLUG_COMMIT",
)


def _detect_git_sha() -> Optional[str]:
    global _cached_git_sha, _git_sha_resolved
    if _git_sha_resolved:
        return _cached_git_sha

    for var in _GIT_SHA_ENV_VARS:
        value = os.environ.get(var)
        if value:
            _cached_git_sha = value
            _git_sha_resolved = True
            return _cached_git_sha

    try:
        result = subprocess.run(
            ["git", "rev-parse", "HEAD"],
            cwd=os.getcwd(),
            capture_output=True,
            text=True,
            timeout=2,
            check=True,
        )
        sha = result.stdout.strip()
        _cached_git_sha = sha or None
    except Exception:
        _cached_git_sha = None

    _git_sha_resolved = True
    return _cached_git_sha


def base_resource_attributes() -> Dict[str, Any]:
    attrs: Dict[str, Any] = {
        "hostname": socket.gethostname(),
        "pid": os.getpid(),
        "python_version": platform.python_version(),
    }
    git_sha = _detect_git_sha()
    if git_sha:
        attrs["git_sha"] = git_sha
    return attrs


def resolve_resource_attributes(
    custom: Optional[ResourceAttributesInput] = None,
) -> Dict[str, Any]:
    base = base_resource_attributes()
    if not custom:
        return base
    extra = custom() if callable(custom) else custom
    merged = dict(base)
    merged.update(extra)
    return merged


def _reset_resource_attributes_cache_for_tests() -> None:
    global _cached_git_sha, _git_sha_resolved
    _cached_git_sha = None
    _git_sha_resolved = False
